networkit/cpp/generators/WattsStrogatzGenerator.cpp

```diff
@@ -1,54 +1,65 @@
 /*
-* WattsStrogatzGenerator.cpp
-*
-*  Created on: 09.07.2014
-*      Author: Simon Bischof
-*/
+ * WattsStrogatzGenerator.cpp
+ *
+ *  Created on: 09.07.2014
+ *      Author: Simon Bischof
+ */
 
+// networkit-format
 #include <networkit/auxiliary/Random.hpp>
 #include <networkit/generators/RegularRingLatticeGenerator.hpp>
 #include <networkit/generators/WattsStrogatzGenerator.hpp>
+#include <networkit/graph/GraphTools.hpp>
 
 namespace NetworKit {
 
 WattsStrogatzGenerator::WattsStrogatzGenerator(count nNodes, count nNeighbors, double p)
-    : nNodes(nNodes), nNeighbors(nNeighbors), p(p) {}
+    : nNodes(nNodes), nNeighbors(nNeighbors), p(p) {
+    if (nNeighbors * 2 >= nNodes - 1) {
+        throw std::runtime_error("nNeighbors*2 cannot be equal to nNodes-1.");
+    }
+    if (nNeighbors >= nNodes / 2 - 1) {
+        nNeighbors = nNodes / 2 - 1;
+    }
+}
 
 Graph WattsStrogatzGenerator::generate() {
-    //generate regular ring lattice as initial graph
-    RegularRingLatticeGenerator R(nNodes, nNeighbors);
-    Graph G = R.generate();
+    // generate regular ring lattice as initial graph
+    auto G = RegularRingLatticeGenerator(nNodes, nNeighbors).generate();
+    std::uniform_int_distribution<node> dist(0, nNodes - 1);
+    std::mt19937_64 generator(Aux::Random::getSeed());
+    std::uniform_real_distribution<double> prob{0.0, std::nexttoward(1.0, 2.0)};
 
-    //list all valid end nodes for rewiring an edge
-    auto validEndNode = [&](node u, node v) {
-        return ! G.hasEdge(u, v) && u != v;
-    };
+    // list all valid end nodes for rewiring an edge
+    auto validEndNode = [&G](const node u, const node v) { return u != v && !G.hasEdge(u, v); };
 
-    //rewire according to Watts and Strogatz model
-    for (node u = 0; u < nNodes; u++) {
+    std::vector<node> prevNeighbors;
+    prevNeighbors.reserve(GraphTools::maxDegree(G));
+
+    // rewire according to Watts and Strogatz model
+    for (node u = 0; u < nNodes; ++u) {
```

Comment:
This loop can use the new graph iterators. Please rebase on Dev and change this loop in
```suggestion
    for (const auto u : G.nodeRange()) {
```